include/rmm/cuda_stream_view.hpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+namespace rmm {
+
+/**
+ * @brief Strongly-typed non-owning wrapper for CUDA streams with default constructor.
+ *
+ * This wrapper is simply a "view": it does not own the lifetime of the stream it wraps.
+ */
+class cuda_stream_view {
+ public:
+  constexpr cuda_stream_view()                        = default;
+  constexpr cuda_stream_view(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view(cuda_stream_view&&)      = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view&&) = default;
+  ~cuda_stream_view()                                       = default;
+
+  // TODO disable construction from 0 after cuDF and others adopt cuda_stream_view
+  // cuda_stream_view(int)            = delete; //< Prevent cast from 0
+  // cuda_stream_view(std::nullptr_t) = delete; //< Prevent cast from nullptr
+  // TODO also disable implicit conversion from cudaStream_t
+
+  /**
+   * @brief Implicit conversion from cudaStream_t.
+   */
+  constexpr cuda_stream_view(cudaStream_t stream) noexcept : stream_{stream} {}
+
+  /**
+   * @brief Get the wrapped stream.
+   *
+   * @return cudaStream_t The wrapped stream.
+   */
+  constexpr cudaStream_t value() const noexcept { return stream_; }
+
+  /**
+   * @brief Implicit conversion to cudaStream_t.
+   */
+  constexpr operator cudaStream_t() const noexcept { return value(); }
```

Comment:
If we're going to make this explicit, I think we should bite the bullet and do it now when we are updating code to use the `cuda_stream_view` type. Otherwise when else are we ever going to do it?

Reply:
I will try in order to estimate the impact on libcudf. I think it  will be large.

Reply:
I think it'll be a bit invasive, but friendly to find/replace. My point is that I don't think it's ever going to be any easier to make this change later than it is now.

Reply:
So this actually isn't that bad. There were several changes to be made within RMM itself, but it turns out libcudf only relies on implicit conversion *from* `cudaStream_t` to `cuda_stream_view`, not the other way around.

Reply:
In fact zero changes required to libcudf. Done.

Reply:
> In fact zero changes required to libcudf. We can convert to `cuda_stream_view` in libcudf asynchronously with this PR.

Any places that call raw `cuda*Async` calls or `kernel<<<..., stream>>>` would need to be updated, but that shouldn't be too bad.

Reply:
No, because cudf Still uses `cudaStream_t` parameters. THe only places that are impacted are uses of RMM classes and APIs, but `cuda_stream_view::cuda_stream_view(cudaStream_t)` makes those all just work without change.  

We will definitely change cuDF to use `cuda_stream_view`, but that doesn't have to happen in sync with this PR, which is nice.